src/main/kotlin/ecommerce/controller/AdminController.kt

```diff
@@ -29,19 +32,35 @@ class AdminController(private val productService: ProductService, private val pr
     @GetMapping("/{id}")
     fun getProductById(
         @PathVariable id: Long,
-    ): Product = productService.findProductById(id)
+        user: AuthenticatedUser,
+    ): Product {
+        if (!user.isAdmin()) {
+            throw ResponseStatusException(HttpStatus.FORBIDDEN, "Admin privileges required")
+        }
+        return productService.findProductById(id)
+    }
 
     @GetMapping("")
     fun getAllProducts(
         @RequestParam(defaultValue = "0")page: Int,
         @RequestParam(defaultValue = "10")size: Int,
         @RequestParam(defaultValue = "name")sortBy: String,
-    ): Page<Product> = productService.findAllProducts(page, size, sortBy)
+        user: AuthenticatedUser,
+    ): Page<Product> {
+        if (!user.isAdmin()) {
+            throw ResponseStatusException(HttpStatus.FORBIDDEN, "Admin privileges required")
+        }
+        return productService.findAllProducts(page, size, sortBy)
+    }
 
     @PostMapping("")
     fun createProduct(
         @Valid @RequestBody productRequest: CreateProductRequest,
+        user: AuthenticatedUser,
     ): ResponseEntity<Product> {
+        if (!user.isAdmin()) {
+            throw ResponseStatusException(HttpStatus.FORBIDDEN, "Admin privileges required")
+        }
         val saved = productService.createProduct(productRequest)
         return ResponseEntity.created(URI.create("/api/products/${saved.id}")).body(saved)
     }
```

Comment:
There are many duplicate validation code in controller.
We could try removing the duplicate code using an interceptor or a filter.
In Spring MVC, what are the differences between an interceptor and a filter, and what roles do they each serve?